File: common/src/commonMain/kotlin/nl/norico/binori/puzzle/CellValue.kt
package nl.norico.binori.puzzle

import androidx.compose.ui.graphics.Color

enum class CellValue(private val index: Int) {
    ZERO(0),
    ONE(1),
    ANY(2);

    private val labels = arrayOf("O", "I", " ")
    private val backgroundColors = arrayOf(Color.Gray, Color.Black, Color.White)
    private val textColors = arrayOf(Color.Black, Color.White, Color.Black)

    override fun toString(): String {
        return labels[this.index]
    }

    fun getBackgroundColor(): Color {
        return backgroundColors[this.index]
    }

    fun getTextColor(): Color {
        return textColors[this.index]
    }

    fun next(): CellValue {
        val allValues = CellValue.values()
        return allValues[(this.index + 1) % allValues.size]
    }
}

File: src/puzzle/cellValue.ts
export enum CellValue {
    ZERO = 0,
    ONE = 1,
    ANY = 2
}

export namespace CellValue {
    const labels = ["O", "I", " "];

    export function toString(value: CellValue): string {
        return labels[value];
    }

    export function next(value: CellValue): CellValue {
        return ((value + 1) % 3) as CellValue;
    }

    export function opposite(value: CellValue): CellValue {
        if (value === CellValue.ZERO) return CellValue.ONE;
        if (value === CellValue.ONE) return CellValue.ZERO;
        return CellValue.ANY;
    }
}
